pkg/querier/tripperware/subquery.go

```diff
@@ -9,35 +9,82 @@ import (
 )
 
 var (
-	ErrSubQueryStepTooSmall = "exceeded maximum resolution of %d points per timeseries in subquery. Try increasing the step size of your subquery"
+	ErrSubQueryStepTooSmall       = "exceeded maximum resolution of %d points per timeseries in subquery, got %d steps. Try increasing the step size of your subquery"
+	ErrSubQueryTotalStepsExceeded = "exceeded total allowed number of subquery steps %d in the query, got %d steps."
 )
 
 const (
 	MaxStep = 11000
 )
 
-// SubQueryStepSizeCheck ensures the query doesn't contain too small step size in subqueries.
-func SubQueryStepSizeCheck(query string, defaultSubQueryInterval time.Duration, maxStep int64) error {
+// SubQueryStepSizeCheck ensures the query doesn't contain too small step size and too many total steps in subqueries.
+func SubQueryStepSizeCheck(query string, defaultSubQueryInterval time.Duration, maxStep, maxTotalSteps int) error {
+	totalSteps, maxSteps := GetSubQueryStepsFromQuery(query, defaultSubQueryInterval)
+	if maxSteps > maxStep {
+		return httpgrpc.Errorf(http.StatusBadRequest, ErrSubQueryStepTooSmall, maxStep, maxSteps)
+	}
+	if totalSteps > maxTotalSteps {
+		return httpgrpc.Errorf(http.StatusBadRequest, ErrSubQueryTotalStepsExceeded, maxTotalSteps, totalSteps)
+	}
+	return nil
+}
+
+// GetSubQueryStepsFromQuery returns total steps of subqueries for the given query.
+func GetSubQueryStepsFromQuery(query string, defaultSubQueryInterval time.Duration) (int, int) {
 	expr, err := parser.ParseExpr(query)
 	if err != nil {
 		// If query fails to parse, we don't throw step size error
 		// but fail query later on querier.
-		return nil
+		return 0, 0
 	}
-	parser.Inspect(expr, func(node parser.Node, nodes []parser.Node) error {
-		e, ok := node.(*parser.SubqueryExpr)
-		if !ok {
-			return nil
-		}
-		step := e.Step
-		if e.Step == 0 {
-			step = defaultSubQueryInterval
+	var maxSteps int
+	totalSteps := traverseBottomUp(nil, &expr, func(parent *parser.Expr, node *parser.Expr) int {
+		e, ok := (*node).(*parser.SubqueryExpr)
+		if ok {
+			step := e.Step
+			if e.Step == 0 {
+				step = defaultSubQueryInterval
+			}
+			iters := int(e.Range / step)
+			if iters > maxSteps {
+				maxSteps = iters
+			}
```

Comment:
Yeah I like the idea and I think about that before. I decided to execute the full query because:
1. The syntax tree traversal should be fast so it should be ok to wait a bit
2. I want to return the exact max total steps and max steps from this function. It is also used as part of the query frontend log to log number of steps each query has.

Reply:
> I want to return the exact max total steps and max steps from this function. It is also used as part of the query frontend log to log number of steps each query has

This is a very good point